samza-core/src/main/scala/org/apache/samza/diagnostics/DiagnosticsStreamMessage.java

```diff
@@ -0,0 +1,305 @@
+  //   * This is typically used when deserializing messages from a diagnostics-stream.
+  //   * @param metricsSnapshot
+  public static DiagnosticsStreamMessage convertToDiagnosticsStreamMessage(MetricsSnapshot metricsSnapshot) {
+    DiagnosticsStreamMessage diagnosticsStreamMessage =
+        new DiagnosticsStreamMessage(metricsSnapshot.getHeader().getJobName(), metricsSnapshot.getHeader().getJobId(),
+            metricsSnapshot.getHeader().getContainerName(), metricsSnapshot.getHeader().getExecEnvironmentContainerId(),
+            metricsSnapshot.getHeader().getVersion(), metricsSnapshot.getHeader().getSamzaVersion(),
+            metricsSnapshot.getHeader().getHost(), metricsSnapshot.getHeader().getTime(),
+            metricsSnapshot.getHeader().getResetTime());
+
+    if (metricsSnapshot.getMetrics().getAsMap().containsKey(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)) {
+      diagnosticsStreamMessage.addContainerNumCores((int) metricsSnapshot.getMetrics()
+          .getAsMap()
+          .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+          .get(CONTAINER_NUM_CORES_METRIC_NAME));
+
+      diagnosticsStreamMessage.addContainerMb((int) metricsSnapshot.getMetrics()
+          .getAsMap()
+          .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+          .get(CONTAINER_MB_METRIC_NAME));
+
+      diagnosticsStreamMessage.addNumStoresWithChangelog((int) metricsSnapshot.getMetrics()
+          .getAsMap()
+          .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+          .get(CONTAINER_NUM_STORES_WITH_CHANGELOG_METRIC_NAME));
+
+      diagnosticsStreamMessage.addContainerModels(deserializeContainerModelMap(
+          (Map<String, Object>) metricsSnapshot.getMetrics()
+              .getAsMap()
+              .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+              .get(CONTAINER_MODELS_METRIC_NAME)));
+
+      diagnosticsStreamMessage.addProcessorStopEvents(
+          (List<DiagnosticsManager.ProcessorStopEvent>) metricsSnapshot.getMetrics()
+              .getAsMap()
+              .get(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)
+              .get(STOP_EVENT_LIST_METRIC_NAME));
+    }
+
+    if (metricsSnapshot.getMetrics().getAsMap().containsKey(SAMZACONTAINER_METRICS_GROUP_NAME)) {
+      diagnosticsStreamMessage.addDiagnosticsExceptionEvents(
+          (Collection<DiagnosticsExceptionEvent>) metricsSnapshot.getMetrics()
+              .getAsMap()
+              .get(SAMZACONTAINER_METRICS_GROUP_NAME)
+              .get(EXCEPTION_LIST_METRIC_NAME));
+    }
+
+    return diagnosticsStreamMessage;
+  }
+
+  /**
+   * Helper method to use {@link SamzaObjectMapper} to serialize {@link ContainerModel}s.
+   * {@link SamzaObjectMapper} provides several conventions and optimizations for serializing containerModels.
+   * @param containerModelMap map of container models to serialize.
+   * @return
+   */
+  private static Map<String, String> serializeContainerModelMap(Map<String, ContainerModel> containerModelMap) {
```

Comment:
Would it be simpler to just return a `String` here? Just serialize the whole `containerModelMap` using `SamzaObjectMapper`.

Reply:
I thought itd be the same since samza-object-mapper and the serde are jackson anyways, so i kept it as a map, where the key count = num of containers.

Reply:
I think it would be the same, but by returning the `String`, the implementation of `serializeContainerModelMap` just becomes `return SamzaObjectMapper.getObjectMapper().writeValueAsString(containerModelMap)`.

Reply:
agree, tried simplifying this.